sourcetracker/_cli/gibbs.py

```diff
@@ -134,96 +131,69 @@ def gibbs(table_fp, mapping_fp, output_dir, loo, jobs, alpha1, alpha2, beta,
-    # shared.
-    o = open(mapping_fp, 'U')
-    sample_metadata_lines = o.readlines()
-    o.close()
+    # Load the metadata file and feature table. Do a data check on the
+    # feature_table.
+    sample_metadata = pd.read_table(mapping_fp, index_col=0)
+    _ft = load_table(table_fp)
+    feature_table = biom_to_df(_ft, cfc=True)
 
-    sample_metadata, biom_table = \
-        _cli_sync_biom_and_sample_metadata(
-            parse_mapping_file(sample_metadata_lines),
-            load_table(table_fp))
+    # Remove samples not shared by both feature and metadata tables and order
+    # rows equivalently.
+    sample_metadata, feature_table = \
+        remove_unshared_samples_and_order(sample_metadata, feature_table)
 
-    # If biom table has fractional counts, it can produce problems in indexing
-    # later on.
-    biom_table.transform(lambda data, id, metadata: np.ceil(data))
-
-    # If biom table has sample metadata, there will be pickling errors when
-    # submitting multiple jobs. We remove the metadata by making a copy of the
-    # table without metadata.
-    biom_table = Table(biom_table._data.toarray(),
-                       biom_table.ids(axis='observation'),
-                       biom_table.ids(axis='sample'))
-
-    # Parse the mapping file and options to get the samples requested for
-    # sources and sinks.
-    source_samples, sink_samples = sinks_and_sources(
-        sample_metadata, column_header=source_sink_column,
-        source_value=source_column_value, sink_value=sink_column_value)
+    # Identify source and sink samples.
+    source_samples = get_samples(sample_metadata, source_sink_column,
+                                 source_column_value)
+    sink_samples = get_samples(sample_metadata, source_sink_column,
+                               sink_column_value)
 
     # If we have no source samples neither normal operation or loo will work.
     # Will also likely get strange errors.
     if len(source_samples) == 0:
-        raise ValueError('Mapping file or biom table passed contain no '
-                         '`source` samples.')
+        raise ValueError('Sample metadata contains no `source` samples under '
+                         'the given `source_sink_column`. Without source '
+                         'samples neither LOO or normal prediction will work.')
 
     # Prepare the 'sources' matrix by collapsing the `source_samples` by their
     # metadata values.
-    sources_envs, sources_data = collapse_sources(source_samples,
-                                                  sample_metadata,
-                                                  source_category_column,
-                                                  biom_table, sort=True)
+    csources = collapse_source_data(sample_metadata, feature_table,
+                                    source_samples, source_category_column,
+                                    'mean')
```

Comment:
The documentation should explicitly state that samples belonging to the same Env, whether they be sinks or sources, are collapsed by taking the mean. Actually, should the method of collapse be allowed to be `random` or `median` or `first`, but set by default to `mean`? Otherwise, users would have to collapse their data and provide a collapsed mapping file.